### ISSUE TITLE
change opacity of all inactive windows.

### ISSUE THREAD

**Issue by vdegenne** [WayfireWM/wayfire#1996]
Unless I'm wrong (and I apologize for that) there is no feature to dim inactive windows.
That would be really helpful for simple-tile.

Personally the only reason I use window borders is because I like having a visual clue of which window is currently selected but if only I could play with the opacity instead I could get rid of the borders and enjoy every pixels of my screen.

It could also be a general solution because some windows can't even have borders so even if we use borders sometimes it's a bit confusing. Using opacity we could determine which window is in use much faster.

**Comment by ammen99** [comment]
You can do that via the IPC. We even have an example of how to do it: https://github.com/WayfireWM/wayfire/blob/master/ipc-scripts/inactive-alpha.py though it is very basic, if you figure out improvements to it, feel free to send a PR.

You need to enable `ipc`, `ipc-rules` and `alpha`, and then run this script either from a terminal or from the autostart plugin.

**Comment by vdegenne** [comment]
I've activated `ipc` and `ipc-rules` and `alpha` but when I run the script I get this error:
```
ModuleNotFoundError: No module named 'wayfire_socket'
```
I tried to install `wayfire-devel` and reboot but I still have this error.  
What am I doing wrong?

**Comment by soreau** [comment]
> I've activated `ipc` and `ipc-rules` and `alpha` but when I run the script I get this error:
> 
> ```
> ModuleNotFoundError: No module named 'wayfire_socket'
> ```
> 
> I tried to install `wayfire-devel` and reboot but I still have this error. What am I doing wrong?

I think you also need wayfire-socket.py from https://github.com/WayfireWM/wayfire/tree/master/ipc-scripts in the same directory where you run the script. Additionally, the script uses WAYFIRE_SOCKET environment variable, which basically means you should load the ipc plugin before autostart, because ipc plugin sets the variable but it won't be picked up by things spawned from autostart (basically the entire shell/session) unless ipc is loaded before autostart.

**Comment by mark-herbert42** [comment]
In fact it would be better to implement something like trailfocus plugin in compiz - not dim all inactive windows, but keep some number of windows not dimmed and start dimming after 2 or 3 (configurable). There can be a case when you need 2-3 windows - like you do write code and have help window open and focus on both of them, or have 2 files with code opened side by side.  So you switch between them and they do not dim, but windows that had focus 3d or later in the list - get dimmed.
Not sure if this possible with wayland thou.

**Comment by soreau** [comment]
> In fact it would be better to implement something like trailfocus plugin in compiz - not dim all inactive windows, but keep some number of windows not dimmed and start dimming after 2 or 3 (configurable). There can be a case when you need 2-3 windows - like you do write code and have help window open and focus on both of them, or have 2 files with code opened side by side. So you switch between them and they do not dim, but windows that had focus 3d or later in the list - get dimmed. Not sure if this possible with wayland thou.

This is definitely possible in wayland, and wayfire makes it relatively easy to write a plugin to do it.

**Comment by ammen99** [comment]
> In fact it would be better to implement something like trailfocus plugin in compiz - not dim all inactive windows, but keep some number of windows not dimmed and start dimming after 2 or 3 (configurable). There can be a case when you need 2-3 windows - like you do write code and have help window open and focus on both of them, or have 2 files with code opened side by side. So you switch between them and they do not dim, but windows that had focus 3d or later in the list - get dimmed. Not sure if this possible with wayland thou.

You can easily extend the script to do that.

**Comment by vdegenne** [comment]
Thanks for y'all help, I'll try today, I'm excited :)

It's not something that you guys are interested in implementing directly as a toggle option in the GUI?

**Comment by ammen99** [comment]
For it to be a toggle option in WCM, we have the following options:

1. Have someone write a C++ plugin to do this, which I find unnecessary for the main repo since we can achieve that with a simple script already
2. Have WCM support enabling such ipc scripts

I don't think either of the two options is ideal, but @soreau might think otherwise (he's the maintainer of wcm and w-p-e).

**Comment by soreau** [comment]
> For it to be a toggle option in WCM, we have the following options:
> 
>     1. Have someone write a C++ plugin to do this, which I find unnecessary for the main repo since we can achieve that with a simple script already
> 
>     2. Have WCM support enabling such ipc scripts
> 
> 
> I don't think either of the two options is ideal, but @soreau might think otherwise (he's the maintainer of wcm and w-p-e).

I think ipc makes it easy to automate tasks and write basic features, but for the expected experience in wcm in its current state, it would need a plugin. However, I would be open to an implementation of a Scripts category in wcm that loads the scripts from a user-owned location as toggle-able plugins, to run or stop them. It would be nice if we could automatically load the required plugins to make the script work, or notify the user of the script plugin requirements somehow.

**Comment by mark-herbert42** [comment]
The point I see - scripts are fine, but is this software developed to be used by programmers only , even more strict - GUI/wayland programmers? Or the idea is to attract more users to it finally and make Wayfire more popular? Because for majority of users including myself to write script doing this is something like rocket science.  This is  really a wekness of Wayfire compared to Compiz - there any "stupid" accountant could define hot corners, window type selection by just clicking on window, build all those logical sequences.  In Wayfire engine is cabable of doing same as compiz could - but without knowing what is window type, window class etc etc  you can configure almost nothing....

**Comment by ammen99** [comment]
First off, you don't have to write the ipc script, it is already written. All you have to do is enable a few plugins and add the script to autostart.

Second, yes, the IPC in its current form is aimed at power users.

Third, nothing prevents interested people from writing the functionality in the form of a C++ plugin, which can be enabled with a single click in WCM. I personally don't use this trailfocus thing, but this is why Wayfire supports plugins - so that other developers can extend it in any way they want. But the plan has never been to write software which has all the functionality maintained by a single person (me).

TLDR: there are no technical constraints and this can be made easy for users. But we need more developers, which are willing to write and maintain plugins.

**Comment by soreau** [comment]
> First off, you don't have to write the ipc script, it is already written. All you have to do is enable a few plugins and add the script to autostart.
> 

I think that trailfocus is more complex than only changing opacity. It is able to change the opacity, brightness and saturation components, based on last focused view. And I am not sure, but I don't think ipc scripts can support this in the current state. See the [trailfocus description here](https://gitlab.com/compiz/compiz-plugins-extra/-/blob/master/metadata/trailfocus.xml.in?ref_type=heads#L5).

**Comment by ammen99** [comment]
Well, adjusting brightness and saturation is not available in any plugin yet anyway .. But yeah, I was not very precise. By trailfocus I meant simply the functionality of changing the inactive windows, currently just alpha.

**Comment by vdegenne** [comment]
It was very easy to install and works great! In that regard yes "scripts are fine" but having a ready to go option would be great.
Why not having the `ipc-scripts` directory installed with Wayfire and if we toggle "dim inactive windows" in the WCM it would move the script in a "enabled" directory in the background? Maybe it's more easy to implement.

**Comment by soreau** [comment]
I have put together an obs plugin and trailfocus script in [this branch of w-p-e](https://github.com/WayfireWM/wayfire-plugins-extra/tree/obs) but the trailfocus script behavior currently relies on [this wayfire patch](http://ix.io/4KXz). To use, apply the patch, enable ipc, stipc, obs and window-rules-ipc plugins and run trailfocus.py in the ipc-scripts/ directory. https://streamable.com/4jikpe

**Comment by soreau** [comment]
For those interested, https://github.com/WayfireWM/wayfire-plugins-extra/commit/d3febdf1499c053c09627b9c9eada89b563560e0

**Comment by NealC** [comment]
Hi,

Links in the top of this thread to

https://github.com/WayfireWM/wayfire/blob/master/ipc-scripts/inactive-alpha.py
https://github.com/WayfireWM/wayfire/tree/master/ipc-scripts

are coming up 404 - do they exist elsewhere?

Many Thanks,
N

**Comment by soreau** [comment]
> Hi,
> 
> Links in the top of this thread to
> 
> https://github.com/WayfireWM/wayfire/blob/master/ipc-scripts/inactive-alpha.py https://github.com/WayfireWM/wayfire/tree/master/ipc-scripts
> 
> are coming up 404 - do they exist elsewhere?
> 
> Many Thanks, N

The scripts have moved to [pywayfire](https://github.com/WayfireWM/pywayfire/tree/main/scripts). You must install pywayfire to use them.

**Comment by NealC** [comment]
Cheers @soreau!
